### PR TITLE
Cardinality reviewer

### PR DESCRIPTION
**Pull request by cmoesel**
This implements the cardinality reviewer, which compares profiles element-by-element to look for relationships between element cardinalities.  See the writeup of expected behavior here: https://github.com/standardhealth/fhir-matches/wiki/Reviewers#cardinality

Note that we currently don't detect cases where a sub-element has been profiled in one resource but not in the other (e.g., `Patient.name.given 1..*` in A but not B), because the other will not have that element at all.  Once we have a package-loader and fisher, we can "unfold" elements to get the original sub-elements and then compare. But we don't have that yet.

We also don't do anything with slices or extensions that exist in one profile but not the other.  I think we may want separate reviewers to look at something like that -- so (at least for now) it is out of scope.

_Since this is the first element-level reviewer, please also think about the way it is structured and if it makes sense -- or if there is room for improvement.  Also, let me know what you think of the CLI since I made changes there._

**To Test:** There are some unit tests, but you can also test out the CLI (and you should, since I made changes there).  Here are a few examples that just use profiles from FHIR R4 Core (so they should be in your FHIR cache). Run them and confirm that the results are correct and make sense:
```
ts-node src/app.ts  -a ~/.fhir/packages/hl7.fhir.r4.core\#4.0.1/package/StructureDefinition-bp.json -b ~/.fhir/packages/hl7.fhir.r4.core\#4.0.1/package/StructureDefinition-bp.json
```
```
ts-node src/app.ts  -a ~/.fhir/packages/hl7.fhir.r4.core\#4.0.1/package/StructureDefinition-bp.json -b ~/.fhir/packages/hl7.fhir.r4.core\#4.0.1/package/StructureDefinition-vitalsigns.json
```
```
ts-node src/app.ts  -a ~/.fhir/packages/hl7.fhir.r4.core\#4.0.1/package/StructureDefinition-vitalsigns.json -b ~/.fhir/packages/hl7.fhir.r4.core\#4.0.1/package/StructureDefinition-bp.json
```
```
ts-node src/app.ts  -a ~/.fhir/packages/hl7.fhir.r4.core\#4.0.1/package/StructureDefinition-bp.json -b ~/.fhir/packages/hl7.fhir.r4.core\#4.0.1/package/StructureDefinition-vitalspanel.json
```
```
ts-node src/app.ts  -a ~/.fhir/packages/hl7.fhir.r4.core\#4.0.1/package/StructureDefinition-bp.json -b ~/.fhir/packages/hl7.fhir.r4.core\#4.0.1/package/StructureDefinition-bmi.json
```

### CODE REVIEW

**Comment by cmoesel** [comment]
Also, don't get too excited, but... I added a box!

File: src/reviewers/sd/CardinalityReviewer.ts [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    // and slices differently,
+    //
+    // TODO: Support profiled subelements once we have a way to look up and
+    //       unfold definitions.
+    // TODO: Consider special logic for extensions and slices.
+    a.snapshot?.element?.forEach(aED => {
+      const bED = b.snapshot?.element?.find(e => e.id === aED.id);
+      if (bED == null) {
+        return;
+      }
```
Comment (mint-thompson) [inline src/reviewers/sd/CardinalityReviewer.ts:23]: There does not appear to be any test that covers the case of an element being present in `a`, but not present in `b`. Could you add a test that covers this?

**Review by mint-thompson — CHANGES_REQUESTED** [review]
Everything seemed in order when run from the command line. I think your approach for an element-level reviewer is good. I have only minor change requests for this.

File: test/reviewers/sd/CardinalityReviewer.test.ts [hunk 2]
```diff
@@ -0,0 +1,259 @@
...
+      expectResult(0, '*', 1, '1', ReviewResult.SUPERSET);
+    });
+
+    it('should assess as superset of B cardinality 1..1', () => {
+      expectResult(0, '*', 2, '3', ReviewResult.SUPERSET);
+    });
+  });
+
+  describe('when A cardinality is 0..1', () => {
+    it('should assess as equivalent to B cardinality 0..*', () => {
```
[the remark this answers is not in the document]
Comment (cmoesel) [reply test/reviewers/sd/CardinalityReviewer.test.ts:131]: Well, darn it.  I'm still waiting for a more intelligent copy/paste.  I think all of our AI research should go into solving this problem.

File: src/reviewers/sd/CardinalityReviewer.ts (hunk 1 above)
Comment (cmoesel) [reply src/reviewers/sd/CardinalityReviewer.ts:23]: Sure thing.  Good idea.

File: test/reviewers/sd/CardinalityReviewer.test.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (cmoesel) [reply test/reviewers/sd/CardinalityReviewer.test.ts]: Addressed in 8acb93a.

File: src/reviewers/sd/CardinalityReviewer.ts (hunk 1 above)
Comment (cmoesel) [reply src/reviewers/sd/CardinalityReviewer.ts:23]: Addressed in 8acb93a.

File: test/reviewers/sd/CardinalityReviewer.test.ts [hunk 3]
```diff
@@ -128,71 +164,71 @@ describe('CardinalityReviewer', () => {
...
       expectResult(0, '0', 0, '0', ReviewResult.EQUIVALENT);
     });
 
-    it('should assess as superset of B cardinality 1..*', () => {
+    it('should assess as disjoint with B cardinality 1..*', () => {
       expectResult(0, '0', 1, '*', ReviewResult.DISJOINT);
     });
 
-    it('should assess as superset of B cardinality 1..1', () => {
+    it('should assess as v B cardinality 1..1', () => {
```
Comment (mint-thompson) [inline test/reviewers/sd/CardinalityReviewer.test.ts:209]: One of the pastes didn't go through correctly:
```suggestion
    it('should assess as disjoint with B cardinality 1..1', () => {
```

Comment (cmoesel) [reply test/reviewers/sd/CardinalityReviewer.test.ts:209]: Oh man.  This is embarrassing.  Once again, I need an AI-enabled VS Code plugin to do this all for me.

What's worse is that I _thought_ I double-checked this all yesterday before pushing.  I guess I should have triple-checked.

Comment (cmoesel) [reply test/reviewers/sd/CardinalityReviewer.test.ts]: All the magic is in 4c465c6.  Although I just noticed I could have easily just committed your suggestion.

I must be off my game today.

**Review by mint-thompson — APPROVED** [review]
hooray for element-level reviews!

**Review by julianxcarter — APPROVED** [review]
Looks good to me!

**Pull request merged**